File: 0568-Maximum Vacation Days/0568-Maximum Vacation Days.cpp
class Solution {
public:
    int maxVacationDays(vector<vector<int>>& flights, vector<vector<int>>& days) {
        int N = flights.size();
        int K = days[0].size();
        
        vector<vector<int>> dp(2, vector<int>(N));
        for (int i = K - 1; i >= 0; --i) {
            for (int j = 0; j < N; ++j) {
                dp[i & 1][j] = dp[(i + 1) & 1][j] + days[j][i];
                for (int k = 0; k < N; ++k) {
                    if (flights[j][k]) {
                        dp[i & 1][j] = max(dp[i & 1][j], dp[(i + 1) & 1][k] + days[k][i]);
                    }                    
                }
            }
        }
        
        return dp[0][0];
    }
};

File: 0501-0600/0568-Maximum Vacation Days/0568-Maximum Vacation Days.py
class Solution:
    def maxVacationDays(self, flights: List[List[int]], days: List[List[int]]) -> int:
        N, K = len(flights), len(days[0])
        dp = [0] * N
        for i in range(K - 1, -1, -1):
            temp = [0] * N
            for j in range(N):
                temp[j] = dp[j] + days[j][i]
                for k in range(N):
                    if flights[j][k]:
                        temp[j] = max(temp[j], dp[k] + days[k][i])
            dp = temp
        return dp[0]
